### ISSUE TITLE
GCC: /lib64 and /usr/lib64 comes before $root/lib when linking

### ISSUE THREAD

**Issue by sebth** [hpcugent/easybuild-easyblocks#921]
When GCC searches for libraries during linking, system `/lib64` and `/usr/lib64` comes before `/lib` of a loaded module. This can cause GCC to link programs against system libraries instead of the libraries of a loaded module.

This is the output of `gcc -print-search-dirs`, which shows the incorrect search order, when GCC-4.8.4 has been installed and loaded (with EasyBuild):

```
install: /home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/
programs: =/home/sebth/eb/software/GCC/4.8.4/libexec/gcc/x86_64-unknown-linux-gnu/4.8.4/:/home/sebth/eb/software/GCC/4.8.4/libexec/gcc/x86_64-unknown-linux-gnu/4.8.4/:/home/sebth/eb/software/GCC/4.8.4/libexec/gcc/x86_64-unknown-linux-gnu/:/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/:/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/:/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/bin/x86_64-unknown-linux-gnu/4.8.4/:/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/bin/
libraries: =
/home/sebth/eb/software/GCC/4.8.4/lib64/x86_64-unknown-linux-gnu/4.8.4/:
/home/sebth/eb/software/GCC/4.8.4/lib64/../lib64/:
/home/sebth/eb/software/GCC/4.8.4/lib/x86_64-unknown-linux-gnu/4.8.4/:
/home/sebth/eb/software/GCC/4.8.4/lib/../lib64/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/lib/x86_64-unknown-linux-gnu/4.8.4/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/lib/../lib64/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../x86_64-unknown-linux-gnu/4.8.4/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../lib64/:
/lib/x86_64-unknown-linux-gnu/4.8.4/:
/lib/../lib64/:
/usr/lib/x86_64-unknown-linux-gnu/4.8.4/:
/usr/lib/../lib64/:
/home/sebth/eb/software/GCC/4.8.4/lib64/:
/home/sebth/eb/software/GCC/4.8.4/lib/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/lib/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../:
/lib/:
/usr/lib/
```

I can think of two ways of solving this.
1. Make sure all packages install 64-bit libraries under `$root/lib64` instead of `$root/lib`. Currently some packages install 64-bit libraries in `$root/lib` and some in `$root/lib64`.
2. Configure GCC to prefer `/lib` over `/lib64`. This can be done with the following change to the GCC EasyBlock:

``` diff
--- easybuild/easyblocks/g/gcc.py.old
+++ easybuild/easyblocks/g/gcc.py
@@ -287,6 +292,18 @@
             self.log.info("Performing regular GCC build...")
             configopts += " --prefix=%(p)s --with-local-prefix=%(p)s" % {'p': self.installdir}

+        # prioritize lib over lib{64,32,x32} for all architectures
+        try:
+            self.log.info("Patching MULTILIB_OSDIRNAMES in gcc/config/i386/t-linux64")
+            src = 'gcc/config/i386/t-linux64'
+            with open(src, 'a') as f:
+                # override default MULTILIB_OSDIRNAMES configuration
+                f.write('MULTILIB_OSDIRNAMES = m64=../lib:../lib64 '
+                        'm32=../lib:../lib32 mx32=../lib:../libx32')
+        except IOError as err:
+            raise EasyBuildError("Failed to patch MULTILIB_OSDIRNAMES in %s: %s",
+                                 src, err)
+
         # III) create obj dir to build in, and change to it
         #     GCC doesn't like to be built in the source dir
         if self.stagedbuild:
```

When using this change `gcc -print-search-dirs` outputs:

```
install: /home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/
programs: =/home/sebth/eb/software/GCC/4.8.4/libexec/gcc/x86_64-unknown-linux-gnu/4.8.4/:/home/sebth/eb/software/GCC/4.8.4/libexec/gcc/x86_64-unknown-linux-gnu/4.8.4/:/home/sebth/eb/software/GCC/4.8.4/libexec/gcc/x86_64-unknown-linux-gnu/:/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/:/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/:/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/bin/x86_64-unknown-linux-gnu/4.8.4/:/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/bin/../lib64/:/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/bin/
libraries: =
/home/sebth/eb/software/GCC/4.8.4/lib64/x86_64-unknown-linux-gnu/4.8.4/:
/home/sebth/eb/software/GCC/4.8.4/lib64/../lib64/:
/home/sebth/eb/software/GCC/4.8.4/lib64/../lib/:
/home/sebth/eb/software/GCC/4.8.4/lib/x86_64-unknown-linux-gnu/4.8.4/:
/home/sebth/eb/software/GCC/4.8.4/lib/../lib64/:
/home/sebth/eb/software/GCC/4.8.4/lib/../lib/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/lib/x86_64-unknown-linux-gnu/4.8.4/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/lib/../lib64/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/lib/../lib/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../x86_64-unknown-linux-gnu/4.8.4/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../lib64/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../lib/:
/lib/x86_64-unknown-linux-gnu/4.8.4/:
/lib/../lib64/:
/lib/../lib/:
/usr/lib/x86_64-unknown-linux-gnu/4.8.4/:
/usr/lib/../lib64/:
/usr/lib/../lib/:
/home/sebth/eb/software/GCC/4.8.4/lib64/:
/home/sebth/eb/software/GCC/4.8.4/lib/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../../x86_64-unknown-linux-gnu/lib/:
/home/sebth/eb/software/GCC/4.8.4/lib/gcc/x86_64-unknown-linux-gnu/4.8.4/../../../:
/lib/:
/usr/lib/
```

If you think choice 2 is a good idea, I can open a pull request for it.

**Comment by boegel** [comment]
@sebth On which OS were you seeing this behaviour?

@geimer Thoughts on this?

**Comment by sebth** [comment]
I am seeing this on CentOS 6.7.

**Comment by boegel** [comment]
@wpoely86 thoughts on this?

**Comment by boegel** [comment]
@wpoely86, @geimer thoughts on this?

**Comment by geimer** [comment]
@boegel Other than wondering whether this is really free of side effects, no specific thoughts...

**Comment by boegel** [comment]
@sebth please take a look at #1030 where I implemented your suggestion, doing this for EasyBuild v3.0 makes a lot of sense w.r.t. backwards compatibility

Also, it may with the RPATH support that is in the works for EasyBuild v3.0 (cfr. https://github.com/hpcugent/easybuild-framework/pull/1942).

**Comment by boegel** [comment]
fixed in #1030, tested extensively and works as intended, so will be included for EasyBuild v3.0, thanks for reporting this and working out a fix @sebth!
